File: src/main/java/org/nia/logic/quests/IQuestStep.java
package org.nia.logic.quests;

import org.nia.model.Quest;
import org.nia.model.QuestEvent;

import java.util.List;
import java.util.Random;

/**
 * @author IANazarov
 */
public interface IQuestStep {

    IQuestEvent getIQuest();

    String getText(Quest quest);

    String getName();

    List<IQuestStep> getNext(Quest quest);

    String getCommand(String formatParam);

    String getGoodText(Quest quest);

    String getBadText(Quest quest);

    default void doWork(QuestEvent questEvent) {}
    default void doFinal(QuestEvent questEvent) {}

    default boolean isWin(QuestEvent questEvent) {
        return new Random().nextInt(100) < questEvent.getWinChance();
    }

    default String getInterceptText() {
        return "";
    }

}

File: src/main/java/ru/nia/tavern/quests/IQuestStep.kt
package ru.nia.tavern.quests

import ru.nia.tavern.model.Quest
import ru.nia.tavern.model.QuestEvent
import java.util.*

/**
 * @author IANazarov
 */
interface IQuestStep {

    val iQuest: IQuestEvent

    val name: String

    val interceptText: String
        get() = ""

    fun getText(quest: Quest): String

    fun getNext(quest: Quest): List<IQuestStep>

    fun getCommand(formatParam: String): String

    fun getGoodText(quest: Quest): String

    fun getBadText(quest: Quest): String

    open fun doWork(questEvent: QuestEvent) {}
    open fun doFinal(questEvent: QuestEvent) {}

    fun isWin(questEvent: QuestEvent): Boolean {
        return Random().nextInt(100) < questEvent.winChance
    }

}
